src/solver.cr

```diff
@@ -0,0 +1,220 @@
+require "./package"
+require "./solver/graph"
+require "./solver/sat"
+require "./spec"
+
+module Shards
+  class Solver
+    setter locks : Array(Dependency)?
+    @solution : Array(Package)?
+
+    def initialize(@spec : Spec, @prereleases = false)
+      @graph = Graph.new
+      @sat = SAT.new
+      @solution = nil
+      @solution_distance = Int32::MAX
+    end
+
+    def prepare(development = true) : Nil
+      @graph.add(@spec, development)
+
+      if locks = @locks
```

Comment:
One scope less with (ignore me if you feel that's too pedantic ;)):

```cr
@locks.try &.each do |lock|
  # ...
end
```

Reply:
Tried it but it's slightly harder to read because of eventual block.